modules/videoio/src/cap_v4l.cpp

```diff
@@ -1585,312 +1533,361 @@ static inline __u32 capPropertyToV4L2(int prop) {
+    v4l2_queryctrl queryctrl = v4l2_queryctrl();
+    queryctrl.id = __u32(v4l2id);
+    if (v4l2id == -1 || !tryIoctl(VIDIOC_QUERYCTRL, &queryctrl)) {
+        fprintf(stderr, "VIDEOIO ERROR: V4L2: property %s is not supported\n", capPropertyName(property_id).c_str());
         return false;
     }
+    _v4l2id = __u32(v4l2id);
+    range = cv::Range(queryctrl.minimum, queryctrl.maximum);
+    if (compatibilityMode) {
+        if (property_id == CV_CAP_PROP_AUTOFOCUS)
+            range = Range(0, 1);
+        else if (property_id == CV_CAP_PROP_AUTO_EXPOSURE)
+            range = Range(0, 4);
+    }
+    return true;
+}
 
-    /* get the min/max values */
-    Range range = capture->getRange(property_id);
-
-    /* scale the value we want to set */
-    value = value * range.size() + range.start;
-
+bool CvCaptureCAM_V4L::icvControl(__u32 v4l2id, int &value, bool isSet) const
+{
     /* set which control we want to set */
-    v4l2_control control = {v4l2id, int(value)};
+    v4l2_control control = v4l2_control();
+    control.id = v4l2id;
+    control.value = value;
 
     /* The driver may clamp the value or return ERANGE, ignored here */
-    if (-1 == ioctl(capture->deviceHandle, VIDIOC_S_CTRL, &control) && errno != ERANGE) {
-        perror ("VIDIOC_S_CTRL");
+    if (!tryIoctl(isSet ? VIDIOC_S_CTRL : VIDIOC_G_CTRL, &control)) {
+        switch (errno) {
+#ifndef NDEBUG
+        case EINVAL:
+            fprintf(stderr,
+                "The struct v4l2_control id is invalid or the value is inappropriate for the given control (i.e. "
+                "if a menu item is selected that is not supported by the driver according to VIDIOC_QUERYMENU).");
+            break;
+        case ERANGE:
+            fprintf(stderr, "The struct v4l2_control value is out of bounds.");
+            break;
+        case EACCES:
+            fprintf(stderr, "Attempt to set a read-only control or to get a write-only control.");
+            break;
+#endif
+        default:
+            perror(isSet ? "VIDIOC_S_CTRL" : "VIDIOC_G_CTRL");
+            break;
+        }
         return false;
     }
+    if (!isSet)
+        value = control.value;
+    return true;
+}
 
-    if(control.id == V4L2_CID_EXPOSURE_AUTO && control.value == V4L2_EXPOSURE_MANUAL) {
-        // update the control range for expose after disabling autoexposure
-        // as it is not read correctly at startup
-        // TODO check this again as it might be fixed with Linux 4.5
-        v4l2_control_range(capture, V4L2_CID_EXPOSURE_ABSOLUTE);
+double CvCaptureCAM_V4L::getProperty(int property_id) const
+{
+    switch (property_id) {
+    case CV_CAP_PROP_FRAME_WIDTH:
+        return form.fmt.pix.width;
+    case CV_CAP_PROP_FRAME_HEIGHT:
+        return form.fmt.pix.height;
+    case CV_CAP_PROP_FOURCC:
+        return palette;
+    case CV_CAP_PROP_FORMAT:
+        return CV_MAKETYPE(IPL2CV_DEPTH(frame.depth), frame.nChannels);
+    case CV_CAP_PROP_MODE:
+        return channelNumber;
```

Comment:
There is some mess with this property (and there was mess before).

For example
- cap_libv4l:
    ```
    - Allow getting/setting CV_CAP_PROP_MODE. These values are supported:
        - CV_CAP_MODE_BGR  : BGR24 (default)
        - CV_CAP_MODE_RGB  : RGB24
        - CV_CAP_MODE_GRAY : Y8, extracted from YUV420
    ```
- DC1394: format presets via `dc1394_video_set_mode()`
- v4l (**patch must be backward compatible**) before this patch: "capture->palette" (like CV_CAP_PROP_FOURCC);
- MSMF: H/W acceleration (???)


There is similar property "for channels" in master branch: `CAP_CROSSBAR_INPIN_TYPE` (with wrong name!)

Lets introduce dedicated property `CAP_PROP_CHANNEL = 43` for that (I will replace `CAP_CROSSBAR_INPIN_TYPE` later).

Reply:
@alalek , I am sorry. I can't understand what do you want from me. Except CV_CAP_PROP_MODE, all "old" parameters used as always and as you says "there was mess before" but - it is **backward compatible** .
cap_libv4l and cap_v4l was not compatible, what should I do with CV_CAP_MODE_GRAY on cap_v4l? MSMF has same CV_CAP_MODE_* as cap_libv4l, but it is not documented. It is very unusual using device capture format as CV_CAP_PROP_MODE. I have device that send me YUYV or MJPEG. To get YUYV frame on windows (msmf) I should set CV_CAP_PROP_MODE to 3 and CV_CAP_PROP_CONVERT_RGB to 1 (WTF!!!), on linux (v4l) I just set CV_CAP_PROP_FOURCC to YUYV (nice!). Why libv4l and MSMF  used CV_CAP_PROP_MODE this way? CV_CAP_PROP_FOURCC - must set device output format.

I used CV_CAP_PROP_MODE for something device/backend special because it was not used for setting a value (v4l).

CAP_PROP_CHANNEL = 43 - one more than Answer to the Ultimate Question of Life, the Universe, and Everything. Wonderful! Should I insert it myself or it was an other PR?

As for me, CV_CAP_PROP_FORMAT (as docs says) is output format of result image, so if you want to get YUV instead of BGR (awful) just set CV_CAP_PROP_FORMAT. For now I configure capture for send me the frame without any transformation and convert it myself

Reply:
Don't touch other backends. Lets process this one.

Actually there are two "processing" formats:
backend SDK -> (1) -> OpenCV -> (2) -> Application

Related properties:
- `CAP_PROP_FOURCC` usually works with (1). V4L stores this value in `capture->palette`
- `CAP_PROP_FORMAT` works with (2) "Format of the %Mat objects returned by VideoCapture::retrieve()"
- `CV_CAP_PROP_CONVERT_RGB` is early version of `CAP_PROP_FORMAT`. Some backends support only one property and don't support another, some of them support both.

(1) is capture/camera "native" format. Like MJPEG from here:
```
$ v4l2-ctl --list-formats
ioctl: VIDIOC_ENUM_FMT
	Index       : 0
	Type        : Video Capture
	Pixel Format: 'YUYV'
	Name        : YUYV 4:2:2

	Index       : 1
	Type        : Video Capture
	Pixel Format: 'MJPG' (compressed)
	Name        : Motion-JPEG
```
`cv::VideoCapture` is able to take this MJPEG stream and then converts it into BGR frames (also it is able to pass it "as is" via `CV_8UC1` buffer).

> cap_libv4l and cap_v4l was not compatible

`cap_libv4l` is going to be deprecated.


> CAP_PROP_CHANNEL = 43

Please add this into this PR and use this property to specify "channelNumber" (old `#define CHANNEL_NUMBER 1`)

---

Please drop `CV_` prefix (it is from legacy C API):
```.diff
-CV_CAP_PROP_...
+CAP_PROP_...
```

Reply:
@alalek I think we really should not care about `CV_CAP_PROP_MODE`

as I described at lengths in #5485, it is just a bad (as in ambiguous) API.

Reply:
@paroj I agree. So lets try avoid using of this ambiguous property.

Reply:
> * `CAP_PROP_FORMAT` works with (2) "Format of the %Mat objects returned by VideoCapture::retrieve()"

Are you sure you want to use the property as described? Because it's useless! I need the output format (after read/retrieve image), and what I get is the number of channels and the size of the component. And that it can be: BGR, RGB, YUV, Lab, HSV, HSL... or not?

Never mind, do as you wish.
I used `CAP_PROP_MODE` for setting backward compatibility mode (by default).
Is that enough to complete my PR?